File: Invalid.java
// Time Complexity :O(length of parenthesis * number of invalid parenthesis)
// Space Complexity :O(length of parenthesis * number of invalid parenthesis)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
class Solution {
    public List<String> removeInvalidParentheses(String s) {
        if(s==null || s.length()==0)
            return new ArrayList<>();
        List<String> output= new ArrayList<>();
        HashSet<String> set = new HashSet<>();//to maintain unique answers
        Queue<String> q = new LinkedList<>();
        boolean found=false;
        q.add(s);
        while(!q.isEmpty()){
            String pop=q.poll();
            if(isValid(pop)){
                found=true;
                output.add(pop);
            }
            if(!found){//we will stop adding to queue if we find the valid parenthesis at a particular level
                for(int i=0;i<pop.length();i++){
                    if(Character.isLetter(pop.charAt(i))){
                        continue;
                    }
                    String s_short=pop.substring(0,i)+pop.substring(i+1,pop.length());//checking all possible string , removing one character at a time
                    if(!set.contains(s_short)){
                        set.add(s_short);
                        q.add(s_short);
                    }
                }
            }
        }
        return output;
    }
    
    
    public boolean isValid(String s){//checks if parenthesis is valid 
        int counter=0;
        for(char ch: s.toCharArray()){
            if(ch=='('){
                counter+=1;
            }
            else if(ch==')'){//if the String starts with ) it means, it can not be valid at all
                if(counter==0)
                    return false;
                counter-=1;
            }
            else//in case of charcters
                continue;
        }
        return counter==0;
    }
}

File: invalid.py
# // Time Complexity : O(2^n)
# // Space Complexity : O(2^n)
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : no

# we do bdf at each level to check if there is a valid string at this level by removing each bracket exhaustively. If a valid is found at a level - we stop- we dont go to create the next level
from Queue import Queue

class Solution(object):
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        # O(2^n), O(2^n)
        def isValid(st):
            count = 0
            for i in range(len(st)):
                c = st[i]
                if c == '(':
                    count += 1
                elif c == ')':
                    count -= 1
                    if count < 0:
                        return False
            return count == 0

        res = []
        hset = set()
        hq = Queue()
        hq.put(s)
        hset.add(s)
        flag = False

        while not hq.empty():
            curr = hq.get()

            if isValid(curr):
                flag = True
                res.append(curr)

            if flag:
                continue  

            for j in range(len(curr)):
                c = curr[j]
                if c.isalpha():
                    continue
                child = curr[:j] + curr[j+1:]
                if child not in hset:
                    hq.put(child)
                    hset.add(child)

        return res


        